File: Sudo/Sources/Sudo/Models/ButtonMode.swift
import Foundation

/// Defines whether a button uses a simple preset action or complex search terms.
enum ButtonMode: Codable, Equatable {
    case simple(SimpleAction)
    case complex

    var isSimple: Bool {
        if case .simple = self { return true }
        return false
    }

    var displayLabel: String {
        switch self {
        case .simple(let action):
            return action.displayName
        case .complex:
            return "Search Terms"
        }
    }
}

File: SudoWindows/Models/ButtonMode.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SudoWindows.Models;

/// <summary>
/// Determines whether a button executes a simple preset action or a complex UI-search action.
/// </summary>
public enum ButtonModeType
{
    Simple,
    Complex
}

/// <summary>
/// Serializable configuration for a single button.
/// </summary>
public class ButtonConfig
{
    [JsonPropertyName("mode")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public ButtonModeType Mode { get; set; } = ButtonModeType.Complex;

    [JsonPropertyName("simpleAction")]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public SimpleActionType? SimpleAction { get; set; }

    [JsonPropertyName("searchTerms")]
    public string[]? SearchTerms { get; set; }

    public ButtonConfig() { }

    public ButtonConfig(ButtonModeType mode, SimpleActionType? simpleAction = null, string[]? searchTerms = null)
    {
        Mode = mode;
        SimpleAction = simpleAction;
        SearchTerms = searchTerms;
    }
}
